src/main/java/com/microsoft/sqlserver/jdbc/SQLServerPreparedStatement.java

```diff
@@ -170,7 +171,13 @@ String getClassNameInternal() {
         cacheKey = new CityHash128Key(sql);
 
         // Parse or fetch SQL metadata from cache.
-        ParsedSQLCacheItem parsedSql = getOrCreateCachedParsedSQL(cacheKey, sql);
+        ParsedSQLCacheItem parsedSql = getCachedParsedSQL(cacheKey, sql);
+        if (null != parsedSql) {
+            isExecutedAtLeastOnce = true;
+        }
+        else {
+            parsedSql = parseAndCacheSQL(cacheKey, sql);
+        }
```

Comment:
> I see why your test would run only sp_executesql with your previous implementation. However, if you kept "already executed once" as part of the handle cache you would not see that behavior, this is why I added that in the first place :-)

I don't understand how that would have made any difference in the *oltpbench* ``epinions`` or ``jpab`` benchmarks (or Hibernate in general), where a single PreparedStatement instance is never executed more than once.  If execution for a single statement never reaches *sp_prepsql* then no handle would be created, right?  The benchmarks below seem to bear this out.

> For the "single call on connection" case the connection level is better, for the other cases I would say difference will be negligible (happy to be proven wrong though :-)

If the difference is negligible, I don't know what we're doing. :grin:  Fortunately, the difference is not negligible.  This is the *oltpbench/epinions* benchmark with the current ``metadataCaching`` branch, with ``enablePrepareOnFirstPreparedStatementCall=false``:
```
Benchmark:     EPINIONS {com.oltpbenchmark.benchmarks.epinions.EpinionsBenchmark}
Configuration: config/sample_epinions_config.xml
Type:          SQLSERVER
Driver:        com.microsoft.sqlserver.jdbc.SQLServerDriver
URL:           jdbc:sqlserver://10.0.40.11:1433;databaseName=epinions;statementPoolingCacheSize=100;serverPreparedStatementDiscardThreshold=100;enablePrepareOnFirstPreparedStatementCall=false;
Isolation:     TRANSACTION_READ_COMMITTED
Scale Factor:  100.0

17:26:53,982 (DBWorkload.java:260) INFO  - ======================================================================
17:26:53,994 (DBWorkload.java:790) INFO  - Creating 10 virtual terminals...
[main] INFO com.zaxxer.hikari.HikariDataSource - HikariPool-1 - Starting...
[main] INFO com.zaxxer.hikari.HikariDataSource - HikariPool-1 - Start completed.
17:26:54,627 (DBWorkload.java:795) INFO  - Launching the EPINIONS Benchmark with 1 Phase...
17:26:54,639 (ThreadBench.java:325) INFO  - PHASE START :: [Workload=EPINIONS] [Serial=false] [Time=120] [Rate=10000] [Arrival=REGULAR] [Ratios=[10.0, 10.0, 10.0, 10.0, 10.0, 10.0, 10.0, 10.0, 20.0]] [ActiveWorkers=10]
17:26:54,640 (ThreadBench.java:472) INFO  - MEASURE :: Warmup complete, starting measurements.
17:28:54,636 (ThreadBench.java:428) INFO  - TERMINATE :: Waiting for all terminals to finish ..
17:28:54,639 (ThreadBench.java:233) INFO  - Starting WatchDogThread
17:28:54,787 (DBWorkload.java:801) INFO  - ======================================================================
17:28:54,787 (DBWorkload.java:802) INFO  - Rate limited reqs/s: Results(nanoSeconds=120000780351, measuredRequests=468917) = 3907.6162557312264 requests/sec
17:28:54,789 (DBWorkload.java:675) INFO  - Output Raw data into file: results/oltpbench.22.csv
17:28:58,977 (DBWorkload.java:613) INFO  - ======================================================================
17:28:58,978 (DBWorkload.java:614) INFO  - Workload Histograms:
GetReviewItemById/01 [37559] ****************************************
GetReviewsByUser/02  [37486] ****************************************
GetAverageRatingB... [37548] ****************************************
GetItemAverageRat... [37485] ****************************************
GetItemReviewsByT... [37568] ****************************************
UpdateUserName/06    [37808] ****************************************
UpdateItemTitle/07   [37365] ***************************************
UpdateReviewRatin... [37629] ****************************************
UpdateTrustRating/09 [74947] ********************************************************************************
```
Again, because no PreparedStatement is ever executed more than once, there are no handles to be shared.  Here is the same test with ``enablePrepareOnFirstPreparedStatementCall=true``:
```
GetReviewItemById/01 [46663] ***************************************
GetReviewsByUser/02  [47003] ****************************************
GetAverageRatingB... [46806] ****************************************
GetItemAverageRat... [47010] ****************************************
GetItemReviewsByT... [47139] ****************************************
UpdateUserName/06    [46886] ****************************************
UpdateItemTitle/07   [46643] ***************************************
UpdateReviewRatin... [47369] ****************************************
UpdateTrustRating/09 [93408] ********************************************************************************
```
And *oltpbench/jpab* (Hibernate) benchmark with ``enablePrepareOnFirstPreparedStatementCall=false``:
```
Persist/01           [ 5641] *****
Retrieve/02          [85171] ********************************************************************************
Update/03            [17168] ****************
Delete/04            [ 5780] *****
```
And with ``enablePrepareOnFirstPreparedStatementCall=true``:
```
Persist/01           [ 6352] *****
Retrieve/02          [92136] ********************************************************************************
Update/03            [18669] ****************
Delete/04            [ 6128] *****
```